### ISSUE TITLE
Ember v4 support

### ISSUE THREAD

**Issue by Windvis** [mu-semtech/ember-data-table#30]
This addon is still used in most projects so making it compatible with Ember 4 is probably a good idea. I know there are plans to give this addon a modern reimplementation but that will take a while (and will probably also change the API) so resolving the deprecations to unblock v4 updates is probably a quicker win for now.

**Comment by madnificent** [comment]
Hi @Windvis.  You're right.  Not fully certain we can remove deprecation warnings without breaking changes.  Also, attention has been given to ... 🥁 ... greater compatibility in feature/enhance-for-ember-4, including some experiments for making consumption of the data table easier.

**Comment by Windvis** [comment]
Ah, I didn't know about that branch :+1:. What kind of deprecations would require a breaking change? I haven't taken a detailed look, but I assumed it was only going to be some `hasBlock` -> `(has-block)` conversion and making sure that everything is prefixed with `this` where needed.

**Comment by Windvis** [comment]
I also realized that we would need to bump ember-auto-import to v2 since that is a requirement for Ember v4. That would be a breaking change since it forces consuming apps to also use v2.

Since we only use it to import lodash's `merge` function it might be better to simply drop both ember-auto-import and lodash, and do the merging in a different way. That can be released in a minor / patch release.

**Comment by madnificent** [comment]
I did not study this in depth as we wanted to revamp the component either case and this seemed like a good time.  To my understanding, consumption is going to be breaking either case.  We currently rely on overriding the templates for the applications that want to extend the ember-data-table.  To my understanding, Ember 4 requires you to import/export the component logic too at that point.  A manual action will therefore need to be taken.  For addons that import and re-export I'm unsure if that is going to be allowed.  I didn't try.

The changes you suggest would also help for the feature/enhance-for-ember-4 branch.  They're general improvements and it'd be nice to have them any case.  If you'd PR, make it a clean commit so we can backport or add it to the feature branch too :D

**Comment by Windvis** [comment]
Overriding templates should still be possible if we keep using `@ember/component`s (and don't colocate the template files _I think_) which should be fine for now to quickly unblock v4 releases?

I would split up the Ember v4 work from the revamp. It could potentially be released as a minor (or major without breaking API changes). The major after that could then be the revamp?
 
Apps aren't then forced to refactor their usage while still being able to update to the 4.4 LTS release once it comes out. I fear that combining this work in 1 major release will leave a lot apps behind on the 3.28 LTS release until they have refactored all the tables to the revamped version which probably won't be high on the priority list :smile:.

I'll try to take a more detailed look and PR it when I find some time :+1:.

**Comment by madnificent** [comment]
Generally pro releasing something that unblocks upgrade to Ember 4.  Keeping an extra variant around has its risks in terms of split and we've been putting off the upgrade for too long.  It could be minimal effort to get something running.

For LBLOD specifically, we should check if the overrides are often used in applications or if they're mostly located in ember-vo-webuniversum-data-table.  We can do so by searching for the handlebars files that can be overriden.  If they're located in ember-vo-webuniversum-data-table then upgrades to Ember 4 should include upgrade of ember-vo-webuniversum-data-table.
